src/psyclone/transformations.py

```diff
@@ -2825,3 +2824,216 @@ def validate(self, loop):
+
+    After applying the transformation the Nodes marked for extraction are \
+    children of the ExtractNode. \
+    Nodes to extract can be individual constructs within an Invoke (e.g. \
+    Loops containing a Kernel or BuiltIn call) or entire Invokes. This \
+    functionality does not support distributed memory.
+    '''
+
+    def __str__(self):
+        return "Extract code within one or more nodes in an invoke."
+
+    @property
+    def name(self):
+        ''' Returns the name of this transformation as a string.'''
+        return "ExtractRegionTrans"
+
+    def _validate(self, node_list):
+        ''' Perform validation checks before applying the transformation
+
+        :param node_list: the list of Node(s) we are checking.
+        :type node_list: list of :py:class:`psyclone.psyGen.Node`.
+        :raises TransformationError: if distributed memory is configured.
+        :raises TransformationError: if transformation is applied to a Node \
+                                     or a list of Nodes which already \
+                                     contain an ExtractNode.
+        :raises TransformationError: if transformation is applied to a \
+                                     Kernel or a BuiltIn call without its \
+                                     parent Loop.
+        :raises TransformationError: if transformation is applied to a Loop \
+                                     without its parent Directive when \
+                                     optimisations are applied.
+        :raises TransformationError: if transformation is applied to an \
+                                     orphaned Directive without its parent \
+                                     Directive.
+        :raises TransformationError: if transformation is applied to a Loop \
+                                     over cells in a colour without its \
+                                     parent Loop over colours in Dynamo0.3 API.
+        :raises TransformationError: if transformation is applied to an \
+                                     inner Loop without its parent outer \
+                                     Loop in GOcean1.0 API.
+        '''
+
+        # First check constraints on Nodes in the node_list common to
+        # all RegionTrans transformations
+        super(ExtractRegionTrans, self)._validate(node_list)
+
+        # Check ExtractRegionTrans specific constraints.
+        from psyclone.extractor import ExtractNode
+        from psyclone.psyGen import Loop, Kern, BuiltIn, Directive, \
+            OMPDoDirective, ACCLoopDirective, OMPParallelDoDirective
+        from psyclone.dynamo0p3 import DynLoop
+        from psyclone.gocean1p0 import GOLoop
+
+        # Extracting distributed memory code is not supported. This
+        # constraint covers the presence of HaloExchange and GlobalSum
+        # classses as they are only generated when distributed memory
+        # is enabled.
+        if Config.get().distributed_memory:
```

Comment:
Removed the constraint as it is now handled by `valid_node_types`. I amodified the former `test_distmem_error` test to just check for `HaloExchange` (now called `test_haloexchange_error`). I also added `test_globalsum_error` instead of the previous check for DistMeme in GOcean 1.0 API (comment above).

Reply:
When I generated code with distributed memory enabled I remembered why I disabled it in the first place. The problem is that class `DynLoop` with distmem on generates infrastructure calls to set halos dirty/clean. I will use the nodes to generate driver layer and there is no simple way to remove the infrastructure calls. Therefore I enabled the restriction again and modified `test_distmem_errors` to check for `HaloExchange` and `GlobalSum` nodes as well.

Reply:
Oh, that's annoying. Never mind, we can always look at this in the future if it becomes a problem.